### ISSUE TITLE
Reflection / Caching / Performance

### ISSUE THREAD

**Issue by benjam-es** [peckphp/peck#22]
## Thoughts
If you go down the route of using reflection, instead of running a class lookup and reflection loop for every checker such as class name checker, method checker, properties checker, annotation checker etc, there may be a more performant way to achieve this goal.

We could introduce the concept of a 'scanner'.

- Classes are loaded and 'reflected into' just once
- We have a cache folder such as .peck
- Each of the things we want to spell check can be then stored in a cache file
    -  Store a filename, line number, type (class / method / property / annotation) and word

We would now have a set of cached scanned words collected after one scanner run, which we can then trigger one or more spell checkers on. We would be simply iteration through json/csv cached words, using their name, line and type for the notifications.

## Benefits

- Reduce repeated code by looping through classes and reflection for each spell checker type
- Single set of 'class scanning' code for future proofing
- Improve performance
- Spell checking becomes a stand alone process using cached words list from scanner output, meaning it is not bound to 'where' words come from

**Comment by c0nst4ntin** [comment]
Interesting proposal! Changes the architecture, but raises good points!
I'll just quickly tag the PRs regarding the Reflection vs. Tokens vs. Regex discussion: #17 #18 #19

**Comment by Jens-dev** [comment]
It would definitely be necessary.
I tested Peck in a small Laravel project because I think the idea is great.
First, I had to ignore around 60 words and 12 folders to get Peck to run, and it still took 1:54 minutes.
Unfortunately, that’s too long.

**Comment by julio-cavallari** [comment]
I created a [first version](https://github.com/peckphp/peck/pull/36) of the cache, but with a slightly different approach than the one suggested.

In this first version, what is cached is only the execution of the spellchecker, no matter where it came from, whether it was an error in the content of a file, in the folder name, or in the file name.

Basically, the checker receives a word, e.g.: `viewersss`, before executing the spellchecker, it checks if the execution of that same word is cached. If there is, it returns the cached suggestions and does not execute the spellchecker again.

With this approach, I was able to reduce the execution time of the tests in the project's main directory from 16.3s to 1.3s.

**Comment by c0nst4ntin** [comment]
As mentioned in the Caching pull request by @julio-cavallari Caching will improve performance on local machines but won't increase performance on ci runners and build pipelines where a new machine/container gets started for each run.

The way it seems in that pull request is that reflection is not the main problem, but the time Aspell takes to check all the words.

**Comment by julio-cavallari** [comment]
Yes, the problem is the spellchecker itself and not reflection

However, the cache will end up improving performance on CI runners as well, as it generates the cache at runtime, the spellchecker will only be executed once per word, so if there are many repeated words, the cache will already reduce the execution time

**Comment by c0nst4ntin** [comment]
Yes in that regard you are totally right. Then problem would only occur when caching the reflection classes I guess?

**Comment by julio-cavallari** [comment]
In fact, the slowness problem occurs due to the amount of words that were checked, when we introduced ClassChecker, the amount of checks made by the cli tool increased significantly.

So the problem is not reflection itself, but rather the amount of words that it brought to be checked.

**Comment by c0nst4ntin** [comment]
Yes exactly, this is what I ment. Sorry if it was a bit unclear

**Comment by GrandadEvans** [comment]
This is just a thought as I haven't looked at the code. but are the words filtered down to a list of unique words

**Comment by chr15k** [comment]
Bit late to this discussion, but after profiling the performance, the primary bottleneck (as mentioned) is the frequency of calls to `aspell->check()` (this accounts for 91% of time it takes to complete the test suite) combined with synchronous execution. It looks like any meaningful first run performance improvements need to be made to the [underlying library](https://github.com/tigitz/php-spellchecker). I did notice these two items on its roadmap which would solve this issue:

- Cache suggestions of already spellchecked words (PSR-6/PSR-16?)
- Add asynchronous mechanism to spellcheckers

**Comment by nunomaduro** [comment]
We are now fast as fuck!
